swiftsimio/accelerated.py

```diff
@@ -168,3 +170,271 @@ def index_dataset(handle: Dataset, mask_array: np.array) -> np.array:
+            chunk_ranges.append([lower, upper])
+
+    return np.asarray(chunk_ranges)
+
+
+@jit(nopython=True, fastmath=True)
+def expand_ranges(ranges: np.ndarray) -> np.array:
+    """
+    Return an array of indices that are within the specified ranges
+
+    Parameters
+    ----------
+    ranges : np.ndarray
+        Array of ranges (see :func:`ranges_from_array`)
+
+    Returns
+    -------
+    np.array
+        1D array of indices that fall within each range specified in `ranges`
+        
+    """
+    length = np.asarray([bounds[1] - bounds[0] for bounds in ranges]).sum()
+
+    output = np.zeros(length, dtype=np.int64)
+    i = 0
+    for bounds in ranges:
+        lower = bounds[0]
+        upper = bounds[1]
+        bound_length = upper - lower
+        output[i : i + bound_length] = np.arange(lower, upper, dtype=np.int64)
+        i += bound_length
+
+    return output
+
+
+@jit(nopython=True, fastmath=True)
+def extract_ranges_from_chunks(
+    array: np.ndarray, chunks: np.ndarray, ranges: np.ndarray
+) -> np.ndarray:
+    """
+    Returns elements from array that are located within specified ranges
+    
+    `array` is a portion of the dataset being read consisting of all the chunks
+    that contain the ranges specified in `ranges`. The `chunks` array contains
+    the indices of the upper and lower bounds of these chunks. To find the 
+    elements of the dataset that lie within the specified ranges we first create
+    an array indexing which chunk each range belongs to. From this information 
+    we create an array of adjusted ranges that takes into account that the array
+    is not the whole dataset. We then return the values in `array` that are 
+    within the adjusted ranges.
+    
+    Parameters
+    ----------
+    array : np.ndarray
+        array containing data read in from snapshot
+    chunks : np.ndarray
+        two dimensional array of bounds for the chunks that contain each range from
+        `ranges`
+    ranges: np.ndarray
+        Array of ranges (see :func:`ranges_from_array`)
+
+    Returns
+    -------
+    np.ndarray
+        subset of `array` whose elements are within each range in `ranges`
+    
+    """
+    # Find out which of the chunks in the chunks array each range in ranges belongs to
+    n_ranges = len(ranges)
+    n_chunks = len(chunks)
+    chunk_array_index = np.zeros(len(ranges), dtype=np.int32)
+    chunk_index = 0
+    i = 0
+    while i < n_ranges and chunk_index < n_chunks:
+        if (
+            chunks[chunk_index][0] <= ranges[i][0]
+            and chunks[chunk_index][1] >= ranges[i][1]
+        ):
+            chunk_array_index[i] = chunk_index
+            i += 1
+        else:
+            chunk_index += 1
+
+
+    # Need to get the locations of the range boundaries with
+    # respect to the indexing in the array of chunked data
+    # (as opposed to the whole dataset)
+    adjusted_ranges = np.copy(ranges)
+    running_sum = 0
+    for i in range(n_ranges-1):
```

Comment:
Here, can we not say `this_chunk = chunks[chunk_array_index[i]]`, and then use this throughout?